### PR TITLE
Contextual matcher

### PR DESCRIPTION
**Pull request by Thomzoy**
Complete overhaul of the `AdvancedRegex` pipe, now named `ContextualMatcher`

- [x] If this PR is a bug fix, the bug is documented in the test suite.
- [x] Changes were documented in the changelog (pending section).
- [x] If necessary, changes were made to the documentation (eg new pipeline).

### CODE REVIEW

File: edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+            max(doclike.start - window, doclike.sent.start) : doclike.end
+        ]
+    elif side == "after":
+        snippet = doclike.doc[
+            doclike.start : min(doclike.end + window, doclike.sent.end)
+        ]
+    return snippet
+
+
+class GroupRegexMatcher(RegexMatcher):
```
Comment (percevalw) [inline edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py:35]: Wouldn't it be better to add a parameter `group_idx` (or something similar) to the RegexMatcher class ?

File: edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        List[Span]
+            List of detected spans.
+        """
+
+        matches = self.phrase_matcher(doc, as_spans=True)
+        regex_matches = self.regex_matcher(doc, as_spans=True)
+
+        spans = list(matches) + list(regex_matches)
+        spans = list(self.filter(spans))
+        spans = list(self.assign(spans))
```
Comment (percevalw) [inline edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py:277]: Do we need these lists ?
Couldn't we instead do:
```suggestion
        spans = (*matches, *regex_matches)
        spans = self.filter(spans)
        spans = self.assign(spans)
```

File: edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+        Get additional information in the context
+        of each entity. This funciton will populate two custom attributes:
+
+        - `ent._.source`
+        - `ent._.assigned`, a dictionary with all retrieved informations
+
+        Parameters
+        ----------
+        spans : List[Span]
+            _description_
```
Comment (percevalw) [inline edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py:337]:
```suggestion
            Spans to enrich
```

File: edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+        Returns
+        -------
+        doc:
+            spaCy Doc object, annotated for extracted terms.
+        """
+
+        ents = self.process(doc)
+
+        doc.spans[self.name] = ents
```
Comment (percevalw) [inline edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py:436]:
```suggestion
        doc.spans[self.name] = list(ents)
```

**Review by percevalw — COMMENTED** [review]
Really nice work, thanks! I've left some comments, and I wonder if the contextual matcher could be tested in a file of its own?

File: edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py (hunk 1 above)
Comment (Thomzoy) [reply edsnlp/pipelines/core/contextual_matcher/contextual_matcher.py]: Yes good thinking, the option was added under the `span_from_group` argument in the `RegexMatcher`

**Review by percevalw — APPROVED** [review]
Excellent work ⭐ the config is much cleaner and more expressive.
I have a few comments and suggestions if I may :)

**Pull request merged**